### PR TITLE
[SC-298] Testy jednostkowe walidatorów

### PR DESCRIPTION
**Pull request by dmncp**
Pora na testy ;)
Instrukcja uruchomienia:
1. Uruchomienie wszystkich testów:
`npm run test`
2. Uruchomienie pojedynczego testu:
* w Webstormie mamy ikonkę w ide, nic trudnego
* w VSC trzeba doinstalować sobie rozszerzenie: [LINK](https://marketplace.visualstudio.com/items?itemName=firsttris.vscode-jest-runner)

Przy okazji znalazłem kilka problemów w walidatorach ;)

### CODE REVIEW

File: frontend/src/tests/unit/validators.test.js [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+    // then
+    expect(validatorMessage1).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage2).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage3).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage4).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage5).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage6).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage7).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage8).toBe(INCORRECT_EMAIL)
```
Comment (kwdrt) [inline frontend/src/tests/unit/validators.test.js]: Okej, to trzeba zrobić refactor - zamiast rozpisywać każdy pojedynczy case tak, można wykorzystać `it.each` po rozbiciu na przypadki dla studenta i prowadzącego do zmiennej z tablicą stringów. Wtedy caaały ten test zwinie się do:

```ts
it.each(invalidEmailsStudent)(
    'should return INCORRECT_EMAIL error for bad email string for student email %s',
    (email) => {
      // when
      const validatorMessage = validateEmail(email, AccountType.STUDENT)

      // then
      expect(validatorMessage).toBe(INCORRECT_EMAIL)
    }
  )
```

(no i tablicy stringów na górze plików.)

Wtedy nawet mamy to rozbite na pojedyncze przypadki w konsoli, więc w razie czego można łatwo zobaczyć na czym się sypnęło:

Code_eCkItwq95s


W zasadzie to nie jest chyba jedyny test który uda się tak zwinąć, i myślę że warto - wtedy dodanie case do dowolnego testu to kwestia dopisania stringa do tablicy zamiast trzech linijek

**Review by kwdrt — CHANGES_REQUESTED** [review]
Myślę że najpierw i tak przejdzie refactor, ale pomyślę o edge case w międzyczasie

File: frontend/src/tests/unit/validators.test.js [hunk 2]
```diff
@@ -100,55 +112,28 @@ describe('Email validator tests:', () => {
...
-    const validatorMessage1 = validateEmail(exampleEmail1, AccountType.STUDENT)
-    const validatorMessage2 = validateEmail(exampleEmail2, AccountType.PROFESSOR)
-    const validatorMessage3 = validateEmail(exampleEmail3, AccountType.PROFESSOR)
-    const validatorMessage4 = validateEmail(exampleEmail4, AccountType.STUDENT)
-    const validatorMessage5 = validateEmail(exampleEmail5, AccountType.PROFESSOR)
-    const validatorMessage6 = validateEmail(exampleEmail6, AccountType.STUDENT)
-    const validatorMessage7 = validateEmail(exampleEmail7, AccountType.PROFESSOR)
-    const validatorMessage8 = validateEmail(exampleEmail8, AccountType.STUDENT)
+    const validatorMessage1 = validateEmail(email, AccountType.STUDENT)
+    const validatorMessage2 = validateEmail(email, AccountType.STUDENT)
```
Comment (kwdrt) [inline frontend/src/tests/unit/validators.test.js]: Czy tutaj 2 razy nie wywołuje się teraz to samo?

File: frontend/src/tests/unit/validators.test.js (hunk 1 above)
Comment (dmncp) [reply frontend/src/tests/unit/validators.test.js]: Dzięki, nie znałem tego it.each :D
Pora się bardziej zagłębić w dokumentację jesta 😅

File: frontend/src/tests/unit/validators.test.js [hunk 3]
```diff
@@ -100,55 +112,28 @@ describe('Email validator tests:', () => {
...
-    const exampleEmail3 = 'z1@agh.edu.pl'
-    const exampleEmail4 = 'F2@student.agh.edu.pl'
+  it.each(correctStudentEmails)('should validate student email %s correctly', (email) => {
+    // when
+    const validatorMessage = validateEmail(email, AccountType.STUDENT)
 
+    // then
+    expect(validatorMessage).toBe('')
+  })
+  it.each(correctProfessorEmails)('should validate student email %s correctly', (email) => {
```
Comment (kwdrt) [inline frontend/src/tests/unit/validators.test.js]: Treść wiadomości nie pokrywa się z case'm:
```suggestion
  it.each(correctProfessorEmails)('should validate professor email %s correctly', (email) => {
```

**Review by kwdrt — APPROVED** [review]
LGTM, ale trzeba poprawić to że 2 razy wywołujemy ten sam test i treść jednego z komunikatów - jest w komentarzach

File: frontend/src/tests/unit/validators.test.js (hunk 2 above)
Comment (dmncp) [reply frontend/src/tests/unit/validators.test.js]: miało być w drugim PROFESSOR, mój błąd :/

File: frontend/src/tests/unit/validators.test.js (hunk 3 above)
Comment (dmncp) [reply frontend/src/tests/unit/validators.test.js]: racja, nie wiem czemu to przegapiłem :/
